# Output data from TractaViewer mining

A data dictionary including sources/calculations for data produced by TractaViewer

### External data input

The list of potential drug target genes, as supplied by the user.

| Field name | Description | Sources |
| --- | --- | --- |
| series | Unique per-gene numeric ID | Generated by TractaViewer |
| HGNC Name | HGNC-recognised symbol - user must supply at least one of this, UniProt ID and Ensembl ID | The HUGO Gene Nomenclature Committee |
| Uniprot ID | UniProt identifier - user must supply at least one of this, HGNC name and Ensembl ID | Uniprot |
| GeneID | Ensembl ID - user must supply at least one of this, UniProt ID and HGNC name | Ensembl |
| Other columns | Other columns (e.g., lab-specific data) which are not used by TractaViewer | User |

### Basic information

A high-level overview of the input genes, including relevant IDs, general function, and homologs in model species.

All other tables include relevant identifiers from this table as appropriate. 

| Field name | Description | Sources |
| --- | --- | --- |
| series | Unique per-gene numeric ID | Generated by TractaViewer |
| HGNC Name | HGNC-recognised symbol | The HUGO Gene Nomenclature Committee |
| GeneID | Ensembl ID | Ensembl |
| Approved Name | Full human-readable gene name | GeneNames |
| Previous Name | Older versions of full human-readable gene name, if any | GeneNames |
| Previous Symbols | Older versions of HGNC symbol, if any | GeneNames |
| Synonyms | Current alternate symbols, if any | GeneNames |
| Uniprot ID | UniProt identifier | Uniprot |
| Entrez ID | ID number from Entrez database | NCBI Entrez |
| Chromosome | Chromosome on which the gene is found in humans | GeneNames |
| Gene family | Name of the gene family to which the gene has been assigned, if any | GeneNames |
| Gene family ID | Numeric gene family ID | GeneNames |
| Functional summary | A brief text description of the gene's known functions | NCBI Entrez |
| Mouse Ensembl ID | Ensembl ID of mouse homolog | Jackson Lab |
| Mouse Uniprot ID | UniProt ID of mouse homolog | Jackson Lab |
| MGI Symbol | Jackson Lab MGI database ID of mouse homolog | Jackson Lab |
| HCOP Rat | Homolog of gene in rat | Hcop/Ensembl BioMart |
| HCOP Worm | Homolog of gene in _C. elegans_ | Hcop/Ensembl BioMart |
| HCOP Fly | Homolog of gene in _D. melanogaster_ | Hcop/Ensembl BioMart |
| HCOP Zebrafish | Homolog of gene in zebrafish | Hcop/Ensembl BioMart |
| HCOP Mouse | Homolog of gene in mouse | Hcop/Ensembl BioMart |
| Locus group | Description of gene's locus (typically "protein-coding gene") | GeneNames | 
| Is protein | Level of evidence indicating that locus is a protein-coding gene | UniProt |
| RNA class | Transcript biotype | Ensembl BioMart |
| Top level protein classes | Highest level of hierarchical functional classification | Human Protein Atlas |

### Buckets

A set of high-level outputs indicating general suitability of targets for different therapeutic approaches. For a detailed description of the decision processes involved, see the help pages built into TractaViewer.

| Field name | Description | Sources |
| --- | --- | --- |
| Safety bucket | Classification of suitability based on safety-related features | Existing drug data (OpenTargets, FDA), mouse phenotype data (Jackson Lab), gene essentiality, associations with single-gene disorders (OrphaNet, Human Phenotype Ontology), enrichment in off-target tissues (Human Protein Atlas) |
| SM Druggability bucket | Classification of suitability for small molecule drug design | Existing ligands (ChEMBL, Pharos), structural indications (ChEMBL, CanSAR), protein sequence/structure homology (PDB, all-vs-all BLASTp results), protein functional homology (gene families) |
| Feasibility bucket | Classification of factors contributing to or easing a drug design program | Availability of antibodies (Pharos), assays (NCBI), affected pathways (OpenTargets), animal models (OpenTargets), protein structures (PDB), literature (OpenTargets, Pharos), protein-protein interactions (BioGRID) |
| AB-ability bucket | Classification of suitability for antibody-based drug design based on known subcellular localisation data | VerSeDa, Human Protein Atlas, Matrisome database (Hynes Lab), Cell Surface Protein Atlas |
| New modality bucket | Classification of suitability for drug design using novel methods, e.g., ASOs, siRNAs | Calculated from AB-ability and SM druggability buckets | 
| Tissue engagement bucket | Classification of suitability for small molecule drug design | Calculated from Human Protein Atlas tissue-specific enrichment | 
| Pharos | General level of evidence/current drug development | Pharos |

### Bucket guide

Description of classification levels of each bucket in the previous table.

### GTEX 

RNA expression values on a per-tissue basis taken directly from the GTEX dataset with no further modification.

### Barres mouse

RNA expression values on a per-nervous-system-tissue basis taken directly from the Barres lab mouse dataset with no further modification.

### Barres human

RNA expression values on a per-nervous-system-tissue basis taken directly from the Barres lab mouse dataset with no further modification.

### HPA Enrichment

Classification based on per-tissue RNA expression data, performed by the Human Protein Atlas. 

### Risk factors

An aggregation of data from multiple sources concerning commonly examined factors indicating potential safety concerns with drugs developed against a given target.

| Field name | Description | Sources |
| --- | --- | --- |
| Phenotype risk score | If a target has phenotypes affecting certain regions of the body, calculate a score based upon the number of withdrawn drugs with mechanisms of action in the corresponding tissue | Human Phenotype Ontology / WITHDRAWN database |
| Withdrawn drug ratio | Ratio of withdrawn to non-withdrawn drugs | WITHDRAWN database |
| Mutational cancer driver genes | Is classified as potential inducer of cancer | Human Protein Atlas |
| COSMIC somatic mutations in cancer genes | Is classified as potential inducer of cancer | Human Protein Atlas |
| Housekeeping gene | No significant enriched expression detected in any tissue | Human Protein Atlas |
| HPA crucial tissues | Enriched expression detected in heart muscle, liver, kidney, testis, uterine, ovary, bone marrow or lymph node tissues - the most common points of drug development failure due to toxicity | Human Protein Atlas |
| HPO disease associations | Gene is associated with disease phenotypes in the indicated tissue | Human phenotype Ontology |

### Rare disease associations 

Single-gene disorders associated with a gene. Effectively an OrphaNet query (via OpenTargets).

### Common disease associations

Complex (multi-gene or statistical) disorders associated with a gene. Effectively an OMIM query (via OpenTargets). 

### MGI Mouse

Mouse phenotypes from the Jackson Lab's MouseMine database.

### CanSAR

Protein structure, model-based druggability assessment, compound screening data and GO term associations for a gene from the CanSAR database.

### Protein structures

Details of high-resolution 3D protein structure models from the Protein Data Bank.

### Antitargets

List of proteins with protein sequence similarity (from pre-computed BLASTp) greater than user-set threshold (default 40%). 

### Core fitness

Indications as to whether the gene is indispensable to the immediate continuation of the life of human cells.

| Field name | Description | Sources |
| --- | --- | --- |
| Core fitness gene | Shown essential to survival based on a CRISPR screen experiment | https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4662922/#!po=76.1364
| CRISPR-screened core fitness gene | Shown essential to survival based on a different CRISPR screen experiment | http://www.g3journal.org/content/early/2017/06/26/g3.117.041277.long
| OGEE human essential gene | Shown essential to survival based on CRISPR screen experiments across a number of cell lines | OGEE database |

### SM Druggability

Aggregated data from multiple sources concerning indications of the suitability of genes for development of small-molecule drugs.

| Field name | Description | Sources |
| --- | --- | --- |
| EBI Tractable | Is flagged as "tractable" in ChEMBL | EBI DrugEBIlity database |
| EBI Druggable | Is flagged as "druggable" in ChEMBL | EBI DrugEBIlity database |
| EBI Ensemble | Average score across a range of druggability assessments internal to the DrugEBIlity database | EBI DrugEBIlity database |
| Domain ID | Most druggable domain ID according to DrugEBIlity assessment | EBI DrugEBIlity database |
| PDB | Protein Data Bank ID of model used for above calculations | EBI DrugEBIlity database |
| Is protein | Level of evidence indicating that locus is a protein-coding gene | UniProt |
| Ligand | Number of known interacting ligands | IUPHAR |
| Endogenous ligand | Number of known interacting endogenous ligands | IUPHAR |
| Main location | Primary subcellular localisation | Human Protein Atlas |
| Top level protein classes | Highest level of hierarchical functional classification | Human Protein Atlas |
| Second level protein classes | More specific level of hierarchical functional classification | Human Protein Atlas |
| Third level protein classes | Most specific level of hierarchical functional classification | Human Protein Atlas |
| Membrane proteins predicted by MDM | Predicted to be a membrane protein by MDM method | Human Protein Atlas |
| GPCRHMM predicted membrane proteins | Predicted to be a membrane protein by method focused on identifying GPCRs | Human Protein Atlas |
| # TM segments | Number of transmembrane segments in protein primary structure predicted by MDM | Human Protein Atlas |
| Has withdrawn drug | Gene has drug known to act on it which has been withdrawn | WITHDRAWN database |
| Withdrawn drug list | List of drugs acting on gene which have been withdrawn | WITHDRAWN database |

### AB-ability

Aggregated data from multiple sources indicating suitability of genes for antibody-based therapeutic approaches, based on bioaccessibility of protein localisation.

| Field name | Description | Sources |
| --- | --- | --- |
| GPCR | Gene is a GPCR | Human Protein Atlas |
| Predicted membrane protein | Predicted by any method to be a membrane protein | Human Protein Atlas |
| Predicted secreted protein | Predicted by any method to be a secreted protein | Human Protein Atlas |
| VerSeDa secretome membership | Gene is listed as part of the secretome in VerSeDa | VerSeDa database |
| Extracellular matrix component | Gene is listed as part of the matrisome | Hynes Lab matrisome database |
| Subcellular location summary | Human-readable description of subcellular localisation | Human Protein Atlas |
| Subcellular location verification | Strength of evidence backing subcellular location summary | Human Protein Atlas |
| Main subcellular locations | Primary subcellular localisation | Human Protein Atlas |
| Additional subcellular locations | Secondary subcellular localisation | Human Protein Atlas |
| GOIDs | Associated GO terms | Gene Ontology consortium | 
| Surfaceome membership (human) | Gene is listed as part of the human surfaceome | Cell Surface Protein Atlas |
| Surfaceome membership confidence (human) | Strength of evidence supporting gene to be listed as part of the human surfaceome | Cell Surface Protein Atlas |
| Surfaceome membership (mouse) | Gene is listed as part of the mouse surfaceome | Cell Surface Protein Atlas |
| Surfaceome membership confidence (mouse) | Strength of evidence supporting gene to be listed as part of the mouse surfaceome | Cell Surface Protein Atlas |
| Secreted proteins predicted by MDSEC, SignalP, Phobius, SPOCTOPUS | Gene product predicted to be secreted by various methods | Human Protein Atlas | 
| Membrane proteins predicted by MDM, MEMSAT3, MEMSAT-SVM, Phobius, SCAMPI, SPOCTOPUS, THUMBUP, TMHMM, GPCRHMM | Gene product predicted to be membrane-bound by various methods | Human Protein Atlas |

### Feasibility

Aggregated data largely from OpenTargets describing factors likely to make research on a given gene a more feasible undertaking. 

| Field name | Description | Sources |
| --- | --- | --- |
| Assays |  | NCBI |
| Antibodies | Availability of antibodies against this gene | Pharos |
| Literature | Gene is mentioned by name in scientific publications | OpenTargets |
| Affected pathway | Gene is known to affect a biochemical pathway | OpenTargets |
| Animal model | Gene has experimental animal model | OpenTargets |

### Existing drugs

A list of existing drugs which are known or suspected to have some interaction with the gene in question, mostly collected via OpenTargets. 

| Field name | Description | Sources |
| --- | --- | --- |
| Disease | Name of associated disease phenotype | OMIM/OrphaNet (via OpenTargets) |
| Association score | Strength of overall association between gene and disease phenotype | OpenTargets |
| Drug name | Name of active compound | OpenTargets |
| Molecule type | Molecule class (generally "Small molecule") | OpenTargets |
| Max clinical phase | Furthest clinical trial reached by the active compound | OpenTargets |
| Boxed warning | Whether the FDA has issued a black box safety warning concerning the active compound | OpenTargets |
| Drug ID | ChEMBL compound identifier | OpenTargets |
| Withdrawn | Whether the active compound has been withdrawn from sale | WITHDRAWN database |
| Withdrawn reason | Reason for withdrawal from sale | WITHDRAWN database |

### Drug toxicology

Details of active compound toxic effects created by querying T3DB for existing drugs active against a gene.

### Pharos

High-level overview of current state of research from the NCBI Pharos database.

| Field name | Description | Sources |
| --- | --- | --- |
| Pharos ID | Unique Pharos ID of gene (equivalent to UniProt ID) | Pharos |
| Druggable class | High-level functional classification of gene | Pharos |
| Pharos | General state of clinical development on gene | Pharos |
| ChEMBL drug | Number of drugs known to act on this gene in ChEMBL | Pharos |
| ChEMBL ligand | Number of ligands known to act on this gene in ChEMBL | Pharos |
| ChEMBL low-activity ligand | Number of ligands known to act on this gene (at low activity levels) in ChEMBL | Pharos |
| Pharos literature | Number of publications explicitly mentioning this gene | Pharos |

### GWAS

Results of queries to GWAS Catalog. 
